libbpf-tools/biostacks.bpf.c

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "biostacks.h"
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile char targ_disk[DISK_NAME_LEN] = {};
+const volatile bool targ_ms = false;
+
+struct internal_rqinfo {
+	u64 start_ts;
+	struct rqinfo rqinfo;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct internal_rqinfo);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} rqinfos SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct rqinfo);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static __always_inline bool disk_filtered(const char *disk)
+{
+	int i;
+
+	for (i = 0; targ_disk[i] != '\0' && i < DISK_NAME_LEN; i++) {
+		if (disk[i] != targ_disk[i])
+			return false;
+	}
+	return true;
+}
+
+static __always_inline
+int trace_start(struct internal_rqinfo *i_rqinfo, void *ctx,
+		struct request *rq, u32 pid, u64 ts)
+{
+	i_rqinfo->start_ts = ts;
+	i_rqinfo->rqinfo.pid = pid;
+	i_rqinfo->rqinfo.kern_stack_size =
+		bpf_get_stack(ctx, i_rqinfo->rqinfo.kern_stack,
+			sizeof(i_rqinfo->rqinfo.kern_stack), 0);
+	bpf_get_current_comm(&i_rqinfo->rqinfo.comm,
+			sizeof(&i_rqinfo->rqinfo.comm));
+	bpf_map_update_elem(&rqinfos, &rq, i_rqinfo, 0);
+	return 0;
+}
+
+SEC("fentry/blk_account_io_start")
+int BPF_PROG(fentry__blk_account_io_start, struct request *rq)
+{
+	u32 pid = bpf_get_current_pid_tgid();
+	struct internal_rqinfo i_rqinfo = {};
+	u64 ts = bpf_ktime_get_ns();
```

Comment:
nit: doing extra work before you know it's I/O for the desired disk

actually, simpler and nicer way to do this would be to make trace_start capture pid and timestamp

Reply:
Agree, thanks.